### PR TITLE
CMake: Fix linkage issues in omrcore

### PR DESCRIPTION
**Pull request by dnakamura**
Signed-off-by: Devin Nakamura <devinn@ca.ibm.com>

### CODE REVIEW

**Comment by rwy0717** [comment]
LGTM

File: omr/CMakeLists.txt [hunk 1]
```diff
@@ -63,19 +63,21 @@ target_include_directories(omrcore
 		.
 )
 
+
+target_link_libraries(omrcore
+	# NOTE: Don't link any other libraries to omrcore
+	# Doing so will almost certainly create a circular dependency in OpenJ9
```
Comment (DanHeidinga) [inline omr/CMakeLists.txt]: `in OpenJ9` -> `in dependent projects`?  Or maybe `in consuming projects`?

**Comment by samolisov** [comment]
@dnakamura A comment from #2523. I don't know why linking omrcore with their dependencies create any problems for the downstream project. Could the reason be in hardcoded `gc` in `omr/CMakeLists.txt`? If so, one of the PR's commits changes hardcoded `gc` to `${OMR_GC_LIB}` and customers will be able to put their own gc library there.

What do you think about `${OMR_HOOK_LIB}` and `${OMR_TRACE_LIB}`? Must omrcore not be linked with those libraries too?

**Comment by charliegracie** [comment]
@dnakamura can you provide more details about this change?  If there are circular dependencies or anything like that we need to resolve those.

**Comment by dnakamura** [comment]
@samolisov  The issue i believe technically exists in omr, however cmake will not complain about circular dependencies if all the nodes are static libraries. However for OpenJ9 `${OMR_GC_LIB}` , `${OMR_HOOK_LIB}`, and `${OMR_TRACE_LIB}` are all shared libraries.  The actual error message I'm getting:
```
CMake Error: The inter-target dependency graph contains the following strongly connected component (cycle):
  "omrcore" of type STATIC_LIBRARY
    depends on "omrgc" (weak)
    depends on "j9trc" (weak)
  "omrgc" of type STATIC_LIBRARY
    depends on "omrcore" (weak)
    depends on "j9trc" (weak)
  "j9trc" of type SHARED_LIBRARY
    depends on "omrcore" (weak)
    depends on "omrgc" (weak)
At least one of these targets is not a STATIC_LIBRARY.  Cyclic dependencies are allowed only among static libraries.
```

**Comment by rwy0717** [comment]
As I understand what @dnakamura just said:

J9's tracepoint shared library needs RAS stuff from `omrcore`, but the j9 GC glue needs tracepoints. So you end up with this circular dependency graph (this is just one possible interpretation):

```
j9trc -> core -> gc+glue -> j9trc
```

The circular dependency is ultimately between omrcore and omrgc. CMake will allows circular dependencies between static libraries, but complains once you introduce shared libraries into the mix, as in j9trc. The long term solution will be to break the core library into multiple components #2497.

**Comment by dnakamura** [comment]
So after talking offline with @rwy0717 , we believe to root of the issue is that in the cmake world we are building `omrvmstartup` (ie the  `omr/startup` directory) as part of the `omrcore` library, instead of as a separate library. By doing this, in addition to moving `omrrasinit.c` and `omrtrcinit.c` into the startup directory, we can eliminate the bulk of, if no all, the circular dependency issues here. I am working on putting a PR together now

**Comment by dnakamura** [comment]
Closed since #2533 is the better approach

**Pull request closed without merging**